SimTracker/SiPixelDigitizer/plugins/SiPixelDigitizerAlgorithm.cc

```diff
@@ -1971,3 +2001,452 @@ void SiPixelDigitizerAlgorithm::module_killing_DB(uint32_t detID) {
+//! usable, > 1 if something is wrong (no reweight done).
+// *******************************************************************************************************
+int SiPixelDigitizerAlgorithm::PixelTempRewgt2D(int id_in, int id_rewgt, array_2d& cluster)
+{
+  // Local variables 
+  int i, j, k, l, kclose;
+  int nclusx, nclusy, success;
+  float xsize, ysize, q50i, q100i, q50r, q10r, q100r, xhit2D, yhit2D, qclust, dist2, dmin2;
+  float xy_in[BXM2][BYM2], xy_rewgt[BXM2][BYM2], xy_clust[TXSIZE][TYSIZE];
+  int denx_clust[TXSIZE][TYSIZE], deny_clust[TXSIZE][TYSIZE];
+  int goodWeightsUsed, nearbyWeightsUsed, noWeightsUsed;
+  float cotalpha, cotbeta; 
+  // success = 0 is returned if everthing is OK
+  success = 0;
+    
+  // Copy the array to remember original charges
+  array_2d clust(cluster);
+
+  // Take the pixel dimensions from the 2D template
+  xsize = templ2D.xsize();
+  ysize = templ2D.ysize();
+  
+  // Calculate the track angles
+
+  if (fabs(track[5]) > 0.){
+    cotalpha = track[3]/track[5]; //if track[5] (direction in z) is 0 the hit is not processed by re-weighting
+    cotbeta = track[4]/track[5];
+  } else {
+    std::cout << "Reweighting angle is not good!" << std::endl;
+    return 9; //returned value here indicates that no reweighting was done in this case
+  }
+
+  // The 2-D templates are defined on a shifted coordinate system wrt the 1D templates
+  if(ydouble[0]) {
+    yhit2D = track[1] - cotbeta*track[2] + ysize;
+  } else {
+    yhit2D = track[1] - cotbeta*track[2] + 0.5*ysize;
+  }
+  if(xdouble[0]) {
+    xhit2D = track[0] - cotalpha*track[2] + xsize;
+  } else {
+    xhit2D = track[0] - cotalpha*track[2] + 0.5*xsize;
+  }
+  
+  // Zero the input and output templates
+  for(i=0; i<BYM2; ++i) {
+    for(j=0; j<BXM2; ++j) {
+      xy_in[j][i] = 0.;
+      xy_rewgt[j][i] = 0.;
+    }
+  }
+
+  // Next, interpolate the CMSSW template needed to analyze this cluster     
+
+  if(!templ2D.xytemp(id_in, cotalpha, cotbeta, xhit2D, yhit2D, ydouble, xdouble, xy_in)) {success = 1;}
+  if(success!=0){
+    std::cout << "Could not find a template!" << std::endl;
+  }
+
+  if(PrintTemplates){
+    std::cout << "Template unirrad: " << std::endl;
+    printCluster(xy_in);
+  }
+
+  q50i = templ2D.s50();
+  //q50i = 0;
+  q100i = 2.*q50i;
+
+  // Check that the cluster container is a 13x21 matrix
+  
+  if(cluster.num_dimensions() != 2) {
+    std::cout << "Cluster is not 2-dimensional. Return." << std::endl;
+    return 3;
+  }
+  nclusx = (int)cluster.shape()[0];
+  nclusy = (int)cluster.shape()[1];
+  if(nclusx != TXSIZE || xdouble.size() != TXSIZE) {
+    std::cout << "Sizes in x do not match: nclusx=" << nclusx << "  xdoubleSize=" << xdouble.size() << "  TXSIZE=" << TXSIZE << ". Return." << std::endl;
+    return 4;
+  }
+  if(nclusy != TYSIZE || ydouble.size() != TYSIZE) {
+    std::cout << "Sizes in y do not match. Return." << std::endl;
+    return 5;
+  }
+  
+  // Sum initial charge in the cluster   
+  
```

Comment:
@schuetzepaul , in many lines below if float is used in computations then all constants better to be also float : 0.f, 2.f, 0.44444f.....

Reply:
@civanch Thanks, I corrected as many of those as I could find.

Update will follow